src/torchcodec/_core/CpuDeviceInterface.cpp

```diff
@@ -266,25 +278,97 @@ int CpuDeviceInterface::convertAVFrameToTensorUsingSwScale(
-      linesizes);
-  return resultHeight;
+  if (needsResize) {
+    // Double swscale path: first convert to RGB24 at original resolution,
+    // then resize in RGB24 space. This ensures transforms happen in the
+    // output color space (RGB24) rather than the input color space (YUV).
+
+    // First pass: color conversion (YUV -> RGB24) at original resolution
+    torch::Tensor intermediateTensor = allocateEmptyHWCTensor(
+        FrameDims(avFrame->height, avFrame->width), torch::kCPU);
+
+    uint8_t* intermediatePointers[4] = {
+        intermediateTensor.data_ptr<uint8_t>(), nullptr, nullptr, nullptr};
+    int intermediateWidth = avFrame->width;
+    int intermediateLinesizes[4] = {intermediateWidth * 3, 0, 0, 0};
+
+    int intermediateHeight = sws_scale(
+        swsContext_.get(),
+        avFrame->data,
+        avFrame->linesize,
+        0,
+        avFrame->height,
+        intermediatePointers,
+        intermediateLinesizes);
+
+    TORCH_CHECK(
+        intermediateHeight == avFrame->height,
+        "First swscale pass failed: intermediateHeight != avFrame->height: ",
+        intermediateHeight,
+        " != ",
+        avFrame->height);
+
+    // Second pass: resize in RGB24 space
+    // Use cached swscale context for resizing, similar to the color conversion
+    // context caching above.
+    SwsFrameContext resizeSwsFrameContext(
+        avFrame->width,
+        avFrame->height,
+        AV_PIX_FMT_RGB24,
+        outputDims.width,
+        outputDims.height);
+
+    if (!resizeSwsContext_ ||
+        prevResizeSwsFrameContext_ != resizeSwsFrameContext) {
+      resizeSwsContext_ = createSwsContext(
+          resizeSwsFrameContext,
+          AVCOL_SPC_RGB,
+          /*outputFormat=*/AV_PIX_FMT_RGB24,
+          /*swsFlags=*/swsFlags_);
+      prevResizeSwsFrameContext_ = resizeSwsFrameContext;
+    }
+
+    uint8_t* srcPointers[4] = {
+        intermediateTensor.data_ptr<uint8_t>(), nullptr, nullptr, nullptr};
+    int srcLinesizes[4] = {avFrame->width * 3, 0, 0, 0};
+
+    uint8_t* dstPointers[4] = {
+        outputTensor.data_ptr<uint8_t>(), nullptr, nullptr, nullptr};
+    int expectedOutputWidth = outputTensor.sizes()[1];
+    int dstLinesizes[4] = {expectedOutputWidth * 3, 0, 0, 0};
+
+    int resultHeight = sws_scale(
+        resizeSwsContext_.get(),
+        srcPointers,
+        srcLinesizes,
+        0,
+        avFrame->height,
+        dstPointers,
+        dstLinesizes);
+
+    return resultHeight;
+  } else {
+    // No resize needed, just color conversion
+    uint8_t* pointers[4] = {
+        outputTensor.data_ptr<uint8_t>(), nullptr, nullptr, nullptr};
+    int expectedOutputWidth = outputTensor.sizes()[1];
+    int linesizes[4] = {expectedOutputWidth * 3, 0, 0, 0};
+    int resultHeight = sws_scale(
+        swsContext_.get(),
+        avFrame->data,
+        avFrame->linesize,
+        0,
+        avFrame->height,
+        pointers,
+        linesizes);
+    return resultHeight;
+  }
```

Comment:
I think we can simplify this code by structuring it, roughly, as:
```c++
// do color conversion
result = sws_scale(/* color conversion params */);
if (needsResize) {
  result = sws_scale(/* resize params */);
}
```
Or is there something I'm missing that makes the above structure difficult to do?